File: dev/generate-kata-commands.py
#!/usr/bin/env python3
"""
Generate thin wrapper Kata commands for each skill.

For each skill in kata-staging/skills/:
- Creates a command that invokes Skill("kata-{skill-name}")
- Skips if command already exists
"""

import yaml
import re
from pathlib import Path

def skill_name_to_command_name(skill_name):
    """Convert skill name (gerund form) to command name (imperative form).

    Examples:
        kata-adding-phases → add-phase
        kata-planning-phases → plan-phase
        kata-executing-plans → execute-plan
        kata-verifying-work → verify-work

    Algorithm:
        1. Remove 'kata-' prefix
        2. Convert gerund to imperative (remove -ing suffix)
        3. Singularize plural nouns (phases → phase, plans → plan)
    """
    # Remove kata- prefix
    if skill_name.startswith('kata-'):
        name = skill_name[5:]
    else:
        name = skill_name

    # Split into words
    words = name.split('-')

    # Process each word
    result_words = []
    for word in words:
        # Convert gerund to imperative (remove -ing)
        if word.endswith('ing'):
            # Handle special cases
            if word.endswith('dding'):  # adding → add
                word = word[:-4]
            elif word.endswith('nning'):  # planning → plan
                word = word[:-4]
            elif word.endswith('tting'):  # setting → set
                word = word[:-4]
            else:
                word = word[:-3]  # general case

        # Singularize common plurals
        if word.endswith('es'):
            word = word[:-2]  # phases → phase
        elif word.endswith('s') and not word.endswith('ss'):
            word = word[:-1]  # plans → plan

        result_words.append(word)

    return '-'.join(result_words)

def generate_command(skill_name, skill_description):
    """Generate a thin wrapper command that invokes a skill.

    Args:
        skill_name: The skill name (e.g., "kata-adding-phases")
        skill_description: The skill description

    Returns:
        tuple: (command_name, command_content)
    """
    command_name = skill_name_to_command_name(skill_name)

    # Extract first sentence for command description
    description_parts = skill_description.split('.')
    short_description = description_parts[0] if description_parts else skill_description
    short_description = short_description.strip()

    command_template = f'''---
name: {command_name}
description: {short_description}
argument-hint: <description>
version: 0.1.0
disable-model-invocation: true
allowed-tools:
  - Read
  - Write
  - Bash
---

## Step 1: Parse Context

Arguments: "$ARGUMENTS"

## Step 2: Invoke Skill

Run the following skill:
`Skill("{skill_name}")`
'''

    return command_name, command_template

def main():
    print("=== Kata Command Generator ===\n")

    # Process all skills in kata-staging/skills/
    skills_dir = Path('dev/transform/kata-staging/skills')
    commands_dir = Path('dev/transform/kata-staging/commands/kata')

    if not skills_dir.exists():
        print(f"Error: Skills directory not found at {skills_dir}")
        print("Run the skill conversion step first.")
        return 1

    print(f"Reading skills from: {skills_dir}")
    print(f"Generating commands in: {commands_dir}\n")

    # Create commands directory
    commands_dir.mkdir(parents=True, exist_ok=True)

    generated = 0
    errors = 0

    # Find all skill directories
    skill_dirs = [d for d in skills_dir.iterdir() if d.is_dir()]

    if not skill_dirs:
        print("No skill directories found.")
        return 1

    for skill_dir in sorted(skill_dirs):
        skill_md = skill_dir / 'SKILL.md'

        if not skill_md.exists():
            print(f"  ⚠ Skipping {skill_dir.name}: No SKILL.md found")
            continue

        try:
            # Read skill frontmatter
            with open(skill_md, 'r', encoding='utf-8') as f:
                content = f.read()

            # Parse frontmatter
            if not content.startswith('---'):
                print(f"  ✗ {skill_dir.name}: No frontmatter found")
                errors += 1
                continue

            parts = content.split('---', 2)
            if len(parts) < 3:
                print(f"  ✗ {skill_dir.name}: Invalid frontmatter structure")
                errors += 1
                continue

            frontmatter = yaml.safe_load(parts[1])

            if not isinstance(frontmatter, dict):
                print(f"  ✗ {skill_dir.name}: Frontmatter is not a dictionary")
                errors += 1
                continue

            skill_name = frontmatter.get('name')
            skill_description = frontmatter.get('description', '')

            if not skill_name:
                print(f"  ✗ {skill_dir.name}: No 'name' field in frontmatter")
                errors += 1
                continue

            # Generate command
            command_name, command_content = generate_command(skill_name, skill_description)

            # Write command file
            command_path = commands_dir / f'{command_name}.md'

            with open(command_path, 'w', encoding='utf-8') as f:
                f.write(command_content)

            print(f"  ✓ {skill_name} → {command_name}.md")
            generated += 1

        except Exception as e:
            print(f"  ✗ {skill_dir.name}: Error - {e}")
            errors += 1

    # Display results
    print()
    print("─" * 60)
    print(f"Generated: {generated} commands")
    if errors > 0:
        print(f"Errors:    {errors} skills")
    print(f"Output:    {commands_dir}")
    print("─" * 60)

    return 0 if errors == 0 else 1

if __name__ == "__main__":
    exit(main())

File: dev/generate-kata-commands.js
#!/usr/bin/env node
/**
 * Generate thin wrapper Kata commands for each skill.
 *
 * For each skill in kata-staging/skills/:
 * - Creates a command that invokes Skill("kata-{skill-name}")
 * - Skips if command already exists
 */

import fs from 'fs';
import path from 'path';

const SKILLS_DIR = 'dev/transform/kata-staging/skills';
const COMMANDS_DIR = 'dev/transform/kata-staging/commands/kata';
const GSD_COMMANDS_DIR = 'dev/transform/gsd-source/commands/gsd';

// Mapping from generated command name to GSD command name
// (handles cases where skill→command conversion differs from original GSD command)
const GSD_COMMAND_NAME_MAP = {
  'check-todo': 'check-todos',
  'start-milestone': 'new-milestone',
  'start-new-milestone': 'new-milestone',
  'start-project': 'new-project',
  'update-kata': 'update',
};

/**
 * Parse YAML frontmatter from content.
 * Returns { frontmatter: {key: value}, body: string }
 */
function parseFrontmatter(content) {
  if (!content.startsWith('---')) {
    return { frontmatter: {}, body: content };
  }

  const parts = content.split('---');
  if (parts.length < 3) {
    return { frontmatter: {}, body: content };
  }

  const frontmatterText = parts[1].trim();
  const body = parts.slice(2).join('---');

  // Simple YAML parsing for key: value pairs
  const frontmatter = {};
  for (const line of frontmatterText.split('\n')) {
    const match = line.match(/^(\S+):\s*(.*)$/);
    if (match) {
      const [, key, value] = match;
      // Remove quotes if present
      frontmatter[key] = value.replace(/^["']|["']$/g, '');
    }
  }

  return { frontmatter, body };
}

/**
 * Convert skill name (gerund form) to command name (imperative form).
 * kata-adding-phases → add-phase
 */
function skillNameToCommandName(skillName) {
  // Remove kata- prefix
  let name = skillName.startsWith('kata-') ? skillName.slice(5) : skillName;

  // Handle special cases first
  const specialCases = {
    'showing-whats-new': 'whats-new',
    'providing-help': 'help',
    'tracking-progress': 'progress',
    'executing-quick-tasks': 'quick',
    'configuring-settings': 'settings',
  };

  if (specialCases[name]) {
    return specialCases[name];
  }

  // Gerund to imperative mapping
  const gerundMap = {
    'adding': 'add',
    'planning': 'plan',
    'executing': 'execute',
    'verifying': 'verify',
    'starting': 'start',
    'debugging': 'debug',
    'checking': 'check',
    'completing': 'complete',
    'discussing': 'discuss',
    'inserting': 'insert',
    'listing': 'list',
    'mapping': 'map',
    'pausing': 'pause',
    'removing': 'remove',
    'researching': 'research',
    'resuming': 'resume',
    'setting': 'set',
    'updating': 'update',
    'auditing': 'audit',
    'showing': 'show',
    'providing': 'provide',
    'tracking': 'track',
    'configuring': 'configure',
  };

  // Split into words
  const words = name.split('-');

  // Process each word
  const resultWords = words.map((word, idx) => {
    // Convert gerund to imperative (first word only)
    if (idx === 0 && gerundMap[word]) {
      return gerundMap[word];
    }

    // Singularize plurals (not first word)
    if (idx > 0) {
      if (word === 'phases') return 'phase';
      if (word === 'milestones') return 'milestone';
      if (word === 'todos') return 'todo';
      if (word === 'codebases') return 'codebase';
      if (word === 'tasks') return 'task';
      if (word === 'projects') return 'project';
      if (word === 'profiles') return 'profile';
    }

    return word;
  });

  return resultWords.join('-');
}

/**
 * Generate a thin wrapper command that invokes a skill.
 *
 * Looks up the original GSD command to use its description and argument-hint,
 * falling back to skill description if GSD command doesn't exist.
 */
function generateCommand(skillName, skillDescription) {
  const commandName = skillNameToCommandName(skillName);

  // Default: extract first sentence from skill description
  let description = (skillDescription.split('.')[0] || skillDescription).trim();
  let argumentHint = '<description>';

  // Look up original GSD command for its description and argument-hint
  const gsdCommandName = GSD_COMMAND_NAME_MAP[commandName] || commandName;
  const gsdCommandPath = path.join(GSD_COMMANDS_DIR, `${gsdCommandName}.md`);
  if (fs.existsSync(gsdCommandPath)) {
    const gsdContent = fs.readFileSync(gsdCommandPath, 'utf-8');
    const { frontmatter: gsdFm } = parseFrontmatter(gsdContent);

    if (gsdFm.description) {
      // Replace gsd/GSD with kata/Kata in the description
      description = gsdFm.description
        .replace(/\bgsd\b/g, 'kata')
        .replace(/\bGSD\b/g, 'Kata');
    }
    if (gsdFm['argument-hint']) {
      argumentHint = gsdFm['argument-hint'];
    }
  }

  const commandContent = `---
name: ${commandName}
description: ${description}
argument-hint: ${argumentHint}
version: 0.1.0
disable-model-invocation: true
allowed-tools:
  - Read
  - Write
  - Bash
---

## Step 1: Parse Context

Arguments: "$ARGUMENTS"

## Step 2: Invoke Skill

Run the following skill:
\`Skill("${skillName}")\`
`;

  return { commandName, commandContent };
}

function main() {
  console.log('=== Kata Command Generator ===\n');

  if (!fs.existsSync(SKILLS_DIR)) {
    console.log(`Error: Skills directory not found at ${SKILLS_DIR}`);
    process.exit(1);
  }

  // Create commands directory
  fs.mkdirSync(COMMANDS_DIR, { recursive: true });

  console.log(`Reading skills from: ${SKILLS_DIR}`);
  console.log(`Generating commands in: ${COMMANDS_DIR}\n`);

  let generated = 0;
  let skipped = 0;
  let errors = 0;

  // Find all skill directories
  const skillDirs = fs.readdirSync(SKILLS_DIR, { withFileTypes: true })
    .filter(d => d.isDirectory())
    .map(d => d.name)
    .sort();

  for (const skillDirName of skillDirs) {
    const skillMdPath = path.join(SKILLS_DIR, skillDirName, 'SKILL.md');

    if (!fs.existsSync(skillMdPath)) {
      console.log(`  ⚠ Skipping ${skillDirName}: No SKILL.md found`);
      continue;
    }

    try {
      const content = fs.readFileSync(skillMdPath, 'utf-8');
      const { frontmatter } = parseFrontmatter(content);

      const skillName = frontmatter.name;
      const skillDescription = frontmatter.description || '';

      if (!skillName) {
        console.log(`  ✗ ${skillDirName}: No 'name' field in frontmatter`);
        errors++;
        continue;
      }

      const { commandName, commandContent } = generateCommand(skillName, skillDescription);
      const commandPath = path.join(COMMANDS_DIR, `${commandName}.md`);

      // Check if command already exists
      if (fs.existsSync(commandPath)) {
        console.log(`  - ${skillName} → ${commandName}.md (exists, skipped)`);
        skipped++;
        continue;
      }

      fs.writeFileSync(commandPath, commandContent);
      console.log(`  ✓ ${skillName} → ${commandName}.md`);
      generated++;

    } catch (e) {
      console.log(`  ✗ ${skillDirName}: Error - ${e.message}`);
      errors++;
    }
  }

  console.log();
  console.log('─'.repeat(60));
  console.log(`Generated: ${generated} commands`);
  console.log(`Skipped:   ${skipped} (already exist)`);
  if (errors > 0) {
    console.log(`Errors:    ${errors} skills`);
  }
  console.log(`Output:    ${COMMANDS_DIR}`);
  console.log('─'.repeat(60));

  process.exit(errors > 0 ? 1 : 0);
}

main();
